Arbitrary order quadrature on pyramids

dominiktassilostill opened the pull request:
Follow up to #17576. Relies on new functionalities introduced there. Based on the formals given [here](https://link.springer.com/article/10.1007/s10915-009-9334-9).

peterrum commented on source/base/quadrature_lib.cc:
```diff
@@ -2378,36 +2378,37 @@ QGaussPyramid<dim>::QGaussPyramid(const unsigned int n_points_1D)
-    {
-      const double Q14 = 1.0 / 4.0;
-      const double Q43 = 4.0 / 3.0;
+  // Gauss-Legendre in x and y direction and Gauss-Jacobi in z direction
```
Could you add some references? Best to the class description!

peterrum replied:
Also we need a test.

drwells commented:
I think you mean that other PR depends on this one, right? Can you rebase and fix the conflict?

drwells reviewed (changes requested):
Thanks. I have some comments on how to improve the tests.

dominiktassilostill commented on tests/simplex/qgauss_pyramid_01.cc:
```diff
@@ -0,0 +1,58 @@
+    {
+      monomial_powers[d] += accuracy / dim;
+      sum += accuracy / dim;
+    }
```
It took a while, now the the results of the quadrature are compared to the exact solution. The error is not printed directly but it is only checked if an error is exceeded.

dominiktassilostill commented on tests/simplex/qgauss_pyramid_02.cc:
```diff
@@ -0,0 +1,103 @@
+        deallog << "Weight does not match" << std::endl;
+    }
+
+  deallog << std::endl;
```
This test only compared the values to the old implementation, so printing them out should do the same.

bangerth replied:
Could you add a comment at the top of the test (like many other tests have) that explains in 2-3 lines what the test does? (Same also for the _01 test while you're there.)

bangerth commented:
/rebuild

bangerth commented:
There is one failing test from this PR: https://ci.tjhei.info/job/dealii-serial/job/PR-17578/6/testReport/junit/projectroot.tests.simplex.polynomials_pyramid_01debug/serial/simplex_polynomials_pyramid_01_debug/

@dominiktassilostill @drwells Do you think we can still get this ready for this release?

dominiktassilostill commented:
@bangerth thanks for catching that.

The problem with the test was that the sequence of the quadrature points changed (in the previous version the z-direction was inverted), it is now fixed. By having the output twice, the diff shows that the test gives the same results in another order. The duplicated output is now also removed from the test.

bangerth commented:
Still no luck :-( https://ci.tjhei.info/job/dealii-serial/job/PR-17578/7/testReport/

bangerth commented on include/deal.II/base/quadrature_lib.h:
```diff
@@ -986,6 +986,10 @@ class QGaussWedge : public Quadrature<dim>
 
 /**
  * Integration rule for pyramid entities.
+ * Taken from @cite Bergot2010, the integration rules generates @p (n_points_1d)^3
```
```suggestion
 * Taken from @cite Bergot2010, the integration rules generate @p (n_points_1d)^3
```

bangerth commented on source/base/quadrature_lib.cc:
```diff
@@ -2376,36 +2376,37 @@ QGaussPyramid<dim>::QGaussPyramid(const unsigned int n_points_1D)
-      const double Q14 = 1.0 / 4.0;
-      const double Q43 = 4.0 / 3.0;
+  // Gauss-Legendre in x and y direction and Gauss-Jacobi in z direction
+  QGauss<1> q_gauss(n_points_1D);
```
```suggestion
  const QGauss<1> q_gauss(n_points_1D);
```

bangerth commented on tests/simplex/polynomials_pyramid_01.cc:
```diff
@@ -60,24 +60,6 @@ test(const unsigned int degree)
         deallog << v << ' ';
       deallog << std::endl;
     }
-
-  for (unsigned int i = 0; i < quad.size(); ++i)
-    {
-      poly.evaluate(quad.point(i),
-                    values,
-                    grads,
-                    grad_grads,
-                    third_derivatives,
-                    fourth_derivatives);
-
-      for (auto v : values)
-        deallog << v << ' ';
-      deallog << std::endl;
-
-      for (auto v : grads)
-        deallog << v << ' ';
-      deallog << std::endl;
-    }
```
This is the mistakenly duplicated piece of code, right? I don't see any differences to the code above it. Or am I missing anything?

dominiktassilostill replied:
Yes, it is the same as the code above it. As I had to fix the test anyway I removed it.

dominiktassilostill committed:
Arbitrary order quadrature on pyramids

bangerth commented:
I'll let @drwells merge since he currently has a hold on this PR. I'm ok with it as is.

bangerth commented:
Separately, I was thinking about this PR during the lunch-time bike race today. Just like for triangles, these tensor product-type formulas are not optimal in the number of quadrature points, right?

Also: taking https://en.wikipedia.org/wiki/Gauss%E2%80%93Jacobi_quadrature as the basis, you use alpha=2 and beta=0 because the cross-sectional area of a pyramid decreases like (1-z)^2, right?

dominiktassilostill commented:
> Separately, I was thinking about this PR during the lunch-time bike race today. Just like for triangles, these tensor product-type formulas are not optimal in the number of quadrature points, right?

They are not optimal, using the techniques from [Witherden and Vincent](https://doi.org/10.1016/j.camwa.2015.03.017) should generate rules with less quadrature points. Hardcoding some of these for lower orders would be an option.

> Also: taking https://en.wikipedia.org/wiki/Gauss%E2%80%93Jacobi_quadrature as the basis, you use alpha=2 and beta=0 because the cross-sectional area of a pyramid decreases like (1-z)^2, right?

Yes, as `x` and `y` scale something like `x*(1-z)` and `y*(1-z)`, the overall factor of `(1-z)^2` results from the transformation.

bangerth commented:
Thanks for confirming, @dominiktassilostill! I put that into #19915.
